internal/helpers/secrets.go

```diff
@@ -251,20 +251,46 @@ func SyncSecret(ctx context.Context, client ctrlclient.Client, obj ctrlclient.Ob
 		}
 		labels[k] = v
 	}
-	// add any annotations configured in meta.Destination.Labels
+
+	lastType := dest.Type
 	dest.Data = data
 	dest.Type = secretType
 	dest.SetAnnotations(meta.Destination.Annotations)
 	dest.SetLabels(labels)
 	dest.SetOwnerReferences(references)
 	logger.V(consts.LogLevelTrace).Info("ObjectMeta", "objectMeta", dest.ObjectMeta)
-
 	if exists {
-		logger.V(consts.LogLevelDebug).Info("Updating secret")
-		if err := client.Update(ctx, dest); err != nil {
-			return err
+		// secret type is immutable, so we need to force recreate the secret when the
+		// type changes.
+		if dest.Type != lastType {
+			logger.V(consts.LogLevelDebug).Info("Recreating secret")
+			// unset the labels so that the owner object does not get enqueued on secret
+			// deletion
+			dest.SetLabels(nil)
+			if err := client.Update(ctx, dest); err != nil {
+				return err
+			}
+
+			// delete the secret
+			if err := client.Delete(ctx, dest); err != nil {
+				return err
+			}
+
+			dest.ResourceVersion = ""
+			dest.Generation = 0
+			dest.SetLabels(labels)
```

Comment:
May want to clear UID and CreationTimestamp too? Though I imagine those would both be overwritten by k8s.

Reply:
It seems like those fields get created, so I think you are correct about them being overwritten.

```
 $ diff -u /tmp/before.yaml /tmp/after.yaml
--- /tmp/before.yaml    2024-02-15 14:26:40
+++ /tmp/after.yaml     2024-02-15 14:27:03
@@ -3,7 +3,7 @@
   .dockerconfigjson:  .....
 kind: Secret
 metadata:
-  creationTimestamp: "2024-02-15T19:26:34Z"
+  creationTimestamp: "2024-02-15T19:26:56Z"
   labels:
     app.kubernetes.io/component: secret-sync
     app.kubernetes.io/managed-by: hashicorp-vso
@@ -18,6 +18,6 @@
     kind: HCPVaultSecretsApp
     name: image-pull-secrets
     uid: f054ffbd-a2c4-4f2b-afc5-b32585b098fb
-  resourceVersion: "348003"
-  uid: 2a61b478-a07d-4735-a938-46e8abf38d37
-type: Opaque
+  resourceVersion: "348051"
+  uid: be705697-2d7a-4e95-b363-0a366fa7975d
+type: kubernetes.io/dockerconfigjson
```